fix: remove new lines from the end of contextual text blocks

absorpheus opened the pull request:
**Summary**

Fixes #30 

* Created a utility helper to remove newlines from the end of `contextualText` text blocks
* Added tests

This removes newline characters which may exist at the end of `contextualText` text blocks which are creating unnecessary new lines in Obsidian's editing mode.

This screenshot illustrates the bug in `contextualText` text blocks using the default template:

<img width="2160" alt="contextualText-newline-characters" src="https://github.com/bandantonio/obsidian-apple-books-highlights-plugin/assets/19255312/30881985-39b1-4e26-a1a6-123fbeb7677f">

This is after the fix is applied:

<img width="2160" alt="contextualText-after-fix" src="https://github.com/bandantonio/obsidian-apple-books-highlights-plugin/assets/19255312/2f67835c-8009-4190-b13f-712f5411f38c">

bandantonio commented:
Thank you.
I was keeping this issue in mind for a while, and your PR is a good starting point.
In my highlights, I also found the following cases (depends on a book):

- Trailing spaces - ` along the way. `
- Newlines and tabs
  - `reboot?\n\n`
  - `project.\n\n\n`
  - `simple answers.\n\t\t`
  - `success.\n\t\t\t`
  - `instead.\n\t\t\t\t\n\t\t\t\t\n\t\t\t\t\t`
  - `a book\n\t\t\t\t\t\t\n\t\t\t\t\t\n\t\t\t\t\t\n\t\t\t\t\t\t\n\t\t\t\t\t\t\t`

So I thought about either making this helper more universal, or merging it (if I/we could find an elegant way) with the existing one that works with indentation: [src/methods/aggregateDetails.ts#L51](https://github.com/bandantonio/obsidian-apple-books-highlights-plugin/blob/master/src/methods/aggregateDetails.ts#L51). That's why the issue is still here 😄

absorpheus commented:
Thank you for sharing, I wasn't aware that other cases existed! I'm thinking it might be better to keep these helper functions modular so we can easily test them, however it may become cumbersome to call multiple functions nested within each other. 

We could have a 'parent' helper function that takes in an array of helper functions and the text as parameters and loops through the helper functions array, invoking the helper function with the outputted result from previous returned text each time.

I'll have a think about it. It would be helpful if we could write some tests so we can better understand the edge cases we are attempting to solve.

absorpheus commented:
> * Trailing spaces - `along the way.`
> * Newlines and tabs
>   
>   * `reboot?\n\n`
>   * `project.\n\n\n`
>   * `simple answers.\n\t\t`
>   * `success.\n\t\t\t`
>   * `instead.\n\t\t\t\t\n\t\t\t\t\n\t\t\t\t\t`
>   * `a book\n\t\t\t\t\t\t\n\t\t\t\t\t\n\t\t\t\t\t\n\t\t\t\t\t\t\n\t\t\t\t\t\t\t`

Are these trailing spaces, newlines and tabs appearing only at the end of the highlighted text block (contextualText) or do they appear in other places too? e.g. in between highlighted paragraphs

bandantonio commented:
> Thank you for sharing, I wasn't aware that other cases existed!

No blame at all! I should have created an issue for that.

UPD: Created #30 Let's continue this discussion there and place this PR on hold for some time.

> I'm thinking it might be better to keep these helper functions modular so we can easily test them, however it may become cumbersome to call multiple functions nested within each other.

Yes, that's how I thought initially. Modularity is a good approach both from coding and testing perspective, but eventually makes the code base a bit noisy since you end up having several helper functions doing relatively similar actions. Plus, wrapping text fragments into more than two helpers may worsen code readability.

> We could have a 'parent' helper function that takes in an array of helper functions and the text as parameters and loops through the helper functions array, invoking the helper function with the outputted result from previous returned text each time.

Yeah, that's one of the possible ways.

> I'll have a think about it. It would be helpful if we could write some tests so we can better understand the edge cases we are attempting to solve.

Yes, TDD definitely makes sense.

bandantonio reviewed (changes requested):
Thank you for the updates.
I left one comment regarding tests, please check it.

Also, please make sure to do rebase of all your commits into one.
Thank you.

absorpheus commented on test/preserveNewlineIndentation.spec.ts:
```diff
@@ -0,0 +1,20 @@
+import { describe, expect, test } from 'vitest'
+import { preserveNewlineIndentation } from 'src/utils'
+
+describe('preserveNewlineIndentation', () => {
```
You're welcome and done 👍🏼

Please confirm if the test meets the requirement of proper indentation when text block is wrapped. 

Also let me know if I've placed the templates and mocks in the correct files or if there's a more appropriate place where they should be put.

Thank you

bandantonio replied:
LGTM, thank you.
You can proceed with rebase.

bandantonio commented on src/types.ts:
```diff
@@ -17,13 +17,17 @@ export interface IBookAnnotation {
     ZANNOTATIONCREATIONDATE: number;
     ZANNOTATIONMODIFICATIONDATE: number;
     ZANNOTATIONSTYLE: 0 | 1 | 2 | 3 | 4 | 5;
+    ZANNOTATIONDELETED: 0 | 1;
```
This field is redundant, as the db query takes only the existing annotations
```suggestion
```

bandantonio commented on test/mocks/aggregateWrappedTextBlockData.ts:
```diff
@@ -0,0 +1,20 @@
+  "annotations": [{
+    "chapter": "Another aggregated Introduction",
+    "contextualText": "This is a contextual text for the aggregated hightlight from the Apple iPhone User Guide\ncontaining a new line to test the preservation of indentation",
+    "highlight": `Chapter 1 introduces the terminology and approach\nthat we're going to use throughout this book. It examines what we actually mean by\nwords like reliability, scalability, and maintainability, and how\nwe can try to achieve these goals.`,
```
This highlight isn't consistent with the rest of the content, please update.

bandantonio commented on test/mocks/renderedTemplate.ts:
```diff
@@ -60,3 +60,13 @@ along with a new line to test the preservation of indentation
+export const renderedWrappedTextBlockTemplateMock = `----
+
+> [!QUOTE]
+>  Chapter 1 introduces the terminology and approach
```
Please update the test to use the proper highlight from the `aggregateWrappedTextBlockData.ts`

absorpheus commented:
@bandantonio I've rebased this branch with master and squashed the commits down to one commit. I force-pushed the branch so the branch on GitHub is consistent with my local branch. I think you'll need to fetch this branch again as it's history has been re-written and will be different to your local branch so apologies for the inconvenience.

Please let me know if you're happy with the changes. Thanks
